### PR TITLE
correctly identify `git@github.com:user/repo` specs as a git repo

### PR DESCRIPTION
**Pull request by nlf**
currently these get identified as being the package `git` with the spec as `github.com:user/repo` which we then incorrectly identify as a directory, causing `npm i git@github.com:user/repo` to "work" but do very much the wrong thing.

this ensures that inputs that start with a `git@hostname:` do not get misidentified and fall through to hosted-git-info, which parses them appropriately allowing the installation to succeed.

## References

Fixes https://github.com/npm/cli/issues/2658

### CODE REVIEW

**Review by wraithgar — APPROVED** [review]

File: package-lock.json [hunk 1]
```diff
@@ -2468,7 +2468,102 @@
...
+        "string-length",
+        "string-width",
+        "strip-ansi",
+        "supports-color",
+        "to-fast-properties",
+        "type-fest",
+        "unicode-length",
+        "widest-line",
+        "wrap-ansi",
+        "yoga-layout-prebuilt"
```
Comment (ruyadorno) [inline package-lock.json:2566]: argh! this thing again!!! me and @wraithgar went a little bit crazy trying to figure out what's happening with the cli that it keeps appending these lines to package-lock, I remember @wraithgar managed to track down the exact version of the cli that started adding these but I'm not sure we had a conclusion on wether the package-lock was broken before and fixed now or the other way round 😅

File: npa.js [hunk 2]
```diff
@@ -17,6 +17,7 @@ let os
 const isWindows = process.platform === 'win32' || global.FAKE_WINDOWS
 const hasSlashes = isWindows ? /\\|[/]/ : /[/]/
 const isURL = /^(?:git[+])?[a-z]+:/i
+const isGit = /^git@[^:]+:/i
```
Comment (ruyadorno) [inline npa.js:20]: what about https://www.npmjs.com/package/git ?

Comment (ruyadorno) [reply npa.js:20]: oohh nvm haven't seen the rest of the regex there 😅

File: package-lock.json (hunk 1 above)
Comment (nlf) [reply package-lock.json:2566]: we did have a release go out that fixed some stuff with bundled dependencies, i don't recall which one though. i was thinking that was likely where it changed

File: npa.js (hunk 2 above)
Comment (isaacs) [reply npa.js:20]: Hmm...

Given that there is a package named `git`, you could have `git@npm:fork-of-git`, which seems like it'd match this regexp inappropriately?

Also, what about a git repo that doesn't use `git` as the username?  `isaacs@izs.me:some/repo.git` for example.

**Review by isaacs — CHANGES_REQUESTED** [review]
This seems like it'll work most of the time, but brought up two edge cases that might cause it to do the wrong thing:

- `git@npm:forked-git` An alias of the fork of the `git` package.  False positive.
- `isaacs@izs.me:some/repo.git` A git remote with a different username.  False negative.

Enthusiastic +1 on all the other test changes, though.

File: npa.js (hunk 2 above)
Comment (nlf) [reply npa.js:20]: yeah, it really relies on the `:`, which thankfully is an invalid character for a tag so we won't accidentally think a tag is a git repo. you did just highlight an edge case though, where `git@file:../my-git` or `git@npm:notreallygit` or even `git@github:npm/git` will fail, so still a good catch!

**Comment by nlf** [comment]
> This seems like it'll work most of the time, but brought up two edge cases that might cause it to do the wrong thing:
> 
> * `git@npm:forked-git` An alias of the fork of the `git` package.  False positive.
i can fix this one by requiring that a `.` exists between the `@` and `:`, which should be safe since valid shorthands there will not include a `.` in that position

> * `isaacs@izs.me:some/repo.git` A git remote with a different username.  False negative.
this one would require loosening the regex so it's any string, followed by an `@` followed by any string including at least one `.` followed by a `:` which i _think_ should be safe. any edge cases you can think of that would be a false positive on that rule would be helpful
> 
> Enthusiastic +1 on all the other test changes, though.

**Review by ruyadorno — APPROVED** [review]

File: npa.js [hunk 3]
```diff
@@ -17,6 +17,7 @@ let os
 const isWindows = process.platform === 'win32' || global.FAKE_WINDOWS
 const hasSlashes = isWindows ? /\\|[/]/ : /[/]/
 const isURL = /^(?:git[+])?[a-z]+:/i
+const isGit = /^[^@]+@[^:.]+\.[^:]+:/i
```
Comment (isaacs) [inline npa.js:20]: I'd put a `.+$` at the end of this, just to be sure that something comes after the `:`.
